xarray/tests/test_units.py

```diff
@@ -78,17 +167,19 @@ def array_extract_units(obj):
     if isinstance(obj, (xr.Variable, xr.DataArray, xr.Dataset)):
         obj = obj.data
 
-    try:
-        return obj.units
-    except AttributeError:
-        return None
+    for unit_lib in unit_libs:
+        if isinstance(obj, unit_lib.quantity_type):
+            return unit_lib.get_unit(obj)
+
+    return None
```

Comment:
it would probably make sense to follow the implementation in `pint-xarray`... It's been quite a while since I wrote this (and didn't really look at it ever since), so there's a few subtle bugs here (for example: `xr.Dataset` has no attribute `.data`, and if it has a variable, that will not give you the underlying array)

Reply:
Do you mean this?
https://github.com/xarray-contrib/pint-xarray/blob/554898a0d75fdcd21eabd26acb744a50f182fec0/pint_xarray/conversion.py#L82

Reply:
close, but I meant the higher-level functions `attach_units`, `convert_units`, `extract_units`, and `strip_units` (and also the `*_unit_attributes` equivalent)